File: framework/languages/java/e2e/RegistrationCodec/Server/Main/src/main/java/systems/zlink/e2e/registrationcodec/main/Configuration/ServerOptions.java
package systems.zlink.e2e.registrationcodec.main.Configuration;
import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("e2e")
public record ServerOptions(
    String serverEndpoint,
    String httpEndpoint,
    String logDir) {
    public ServerOptions {
        required(serverEndpoint, "server-endpoint"); required(httpEndpoint, "http-endpoint"); required(logDir, "log-dir");
    }
    private static void required(String value, String name) {
        if (value == null || value.isBlank()) throw new IllegalArgumentException("e2e." + name + " is required");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Server/InvalidDuplicate/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/invalidduplicate/Configuration/ServerOptions.kt
package systems.zlink.e2e.kotlin.registrationcodec.invalidduplicate.configuration

data class ServerOptions(
    val serverEndpoint: String,
    val logDir: String,
) {
    companion object {
        fun parse(args: Array<String>): ServerOptions {
            val values = mutableMapOf<String, MutableList<String>>()
            var index = 0
            while (index < args.size) {
                val key = args[index]
                require(key.startsWith("--")) { "Unexpected argument '$key'." }
                require(index + 1 < args.size) { "Missing value for '$key'." }
                values.getOrPut(key) { mutableListOf() }.add(args[index + 1])
                index += 2
            }

            fun get(name: String): String? = values[name]?.lastOrNull()
            return ServerOptions(
                serverEndpoint = get("--server-endpoint") ?: "",
                logDir = get("--log-dir") ?: "logs",
            )
        }
    }
}
